x/proof/keeper/msg_server_submit_proof.go

```diff
@@ -55,15 +77,84 @@ func (k msgServer) SubmitProof(ctx context.Context, msg *types.MsgSubmitProof) (
 		return nil, status.Error(codes.InvalidArgument, err.Error())
 	}
 
+	// Unmarshal the closest merkle proof from the message.
+	sparseMerkleClosestProof := &smt.SparseMerkleClosestProof{}
+	if err := sparseMerkleClosestProof.Unmarshal(msg.GetProof()); err != nil {
+		return nil, types.ErrProofInvalidProof.Wrapf(
+			"failed to unmarshal closest merkle proof: %s",
+			err,
+		)
+	}
+
+	// Get the relay request and response from the proof.GetClosestMerkleProof.
+	closestValueHash := sparseMerkleClosestProof.ClosestValueHash
+	relayBz := closestValueHash[:len(closestValueHash)-sumSize]
+	relay := &servicetypes.Relay{}
+	if err := k.cdc.Unmarshal(relayBz, relay); err != nil {
+		return nil, types.ErrProofInvalidRelay.Wrapf(
+			"failed to unmarshal relay: %s",
+			err,
+		)
+	}
+
+	if err := relay.GetReq().ValidateBasic(); err != nil {
+		return nil, status.Error(codes.FailedPrecondition, err.Error())
+	}
+
+	if err := relay.GetRes().ValidateBasic(); err != nil {
+		return nil, status.Error(codes.FailedPrecondition, err.Error())
+	}
+
+	// Verify that the relay request session header matches the proof session header.
+	if err := compareSessionHeaders(msg.GetSessionHeader(), relay.GetReq().Meta.GetSessionHeader()); err != nil {
+		return nil, status.Error(codes.FailedPrecondition, err.Error())
+	}
+
+	// Verify that the relay response session header matches the proof session header.
+	if err := compareSessionHeaders(msg.GetSessionHeader(), relay.GetRes().Meta.GetSessionHeader()); err != nil {
+		return nil, status.Error(codes.FailedPrecondition, err.Error())
+	}
+
+	supplierPubKey, err := k.pubKeyClient.GetPubKeyFromAddress(ctx, msg.GetSupplierAddress())
```

Comment:
Let's move the "simple checks" closer to the top so the E2E validation "reads easier"